language-server/src/references.js

```diff
@@ -0,0 +1,117 @@
+import { JsoncInstanceFromJSON } from "./json-instance-from-json.js";
+import { validate } from "./json-schema.js";
+import { JsoncInstance } from "./jsonc-instance.js";
+import { buildDiagnostic, extractBaseUri, isValidUrl } from "./util.js";
+
+// TODO: Handle anchor fragments
+
+/**
+ *
+ * @param {JsoncInstance} instance
+ * @returns
+ */
+export const validateReferences = async (instance) => {
+  const diagnostics = [];
+  const promises = [];
+  let baseUri = "";
+  /**
+   *
+   * @param {JsoncInstance} instance
+   * @param {string} basePath
+   * @returns
+   */
+  async function validateRefs(instance, basePath = "") {
+    if (instance.typeOf() === "object") {
+      for (const [key, valueInstance] of instance.entries()) {
+        if (key.value() == "$id" && typeof valueInstance.value() === "string") {
+          baseUri = extractBaseUri(valueInstance.value());
+        }
+        if (
+          key.value() === "$ref" &&
```

Comment:
The vocabulary system allows you to do a lot unusual things including changing the name of the token used to represent a reference. That means you can't just hardcode to the string `$ref`. To get correct keyword name for a given dialect, use the [`getKeywordName`](https://github.com/hyperjump-io/json-schema?tab=readme-ov-file#api-2) function from `@hyperjump/json-schema`. There are two versions of references using the keywordIds `https://json-schema.org/keyword/ref` and `https://json-schema.org/keyword/draft-04/ref`. For the purposes of this functionality there isn't a difference between the two, but you'll need to check for both.